File: api-interop-layer/util/fetch.js
import undici from "undici";
import {
  getFromRedis,
  getTTLFromResponse,
  saveToRedis,
  USE_REDIS,
} from "../redis.js";
import { logger } from "./monitoring/index.js";
import { sleep } from "./sleep.js";

const fetchLogger = logger.child({ subsystem: "fetch wrapper" });
const redisLogger = logger.child({ subsystem: "redis" });

const DISABLE_REDIS = process.env.DISABLE_REDIS ? true : false;
const BASE_URL = process.env.API_URL ?? "https://api.weather.gov";
const BASE_NWSCONNECT_URL = process.env.NWSCONNECT_API_URL ?? "https://preview-api.weather.gov";

const STANDARD_HEADERS = {
  "User-Agent": "beta.weather.gov interop",
};
if (process.env.API_KEY) {
  STANDARD_HEADERS["API-Key"] = process.env.API_KEY;
}

// Set the maximum number of open connections we will check against
// before any requests are actually sent out. 
export const MAX_OPEN_CONNECTIONS = parseInt(process.env.MAX_OPEN_CONNECTIONS) || 16_000;

const AGENT = new undici.Agent({ pipelining: 10, allowH2: true, connections: 20 });
undici.setGlobalDispatcher(
  AGENT,
);

/**
 * Helper function to get all of the current unresolved connections for every
 * origin that the unidici Agent is tracking
 */
export const getNumOpenConnections = () => {
  const stats = AGENT.stats;
  return Object.keys(stats).map(key => {
    const pool = stats[key];
    return pool.size;
  }).reduce((a, b) => {
    return a + b;
  }, 0);
};

/**
 * Helper function that returns true if the  number of
 * open  unidici connections is at or exceeds
 * the configured max number.
 */
export const atMaxNumConnections = () => {
  const openConnections = getNumOpenConnections();
  fetchLogger.trace({ openConnections }, "current open connections");
  if (openConnections >= MAX_OPEN_CONNECTIONS) {
    return true;
  }
  return false;
};

fetchLogger.trace({ max: MAX_OPEN_CONNECTIONS }, "Initializing max open connections");

const internalFetch = async (path) => {
  let url = URL.canParse(path) ? URL.parse(path) : new URL(path, BASE_URL);

  const headers = { ...STANDARD_HEADERS };

  // If the hostname is specified in the original request, preserve that as
  // a special header. In dev environments, our magic proxy will pick this up
  // to use for routing.
  if (URL.canParse(path)) {
    headers["wx-host"] = url.hostname;
  }

  const isWxStory = url.hostname === url.pathname.endsWith("/weatherstories");

  const isBriefing = url.hostname === url.pathname.endsWith("/briefings");

  if (isWxStory || isBriefing) {
    // If the incoming path matches a request to the website's weather story
    // or briefing endpoint, switch to the NWS Connect URL.
    url = new URL(url.pathname, BASE_NWSCONNECT_URL);
  }

  // Attempt to look up the request in the redis cache.
  // If present, we simply resolve with that response.
  // Otherwise, continue making the full network
  // request.
  // Note that we explicitly do not cache GHWO requests
  // for the moment.
  if (!DISABLE_REDIS && USE_REDIS) {
    url = new URL(url);
    const cachedValue = await getFromRedis(url.pathname);
    if (cachedValue) {
      redisLogger.trace({ pathname: url.pathname }, "returning cached value");
      return new Promise((resolve) => {
        resolve(cachedValue);
      });
    }
  }

  fetchLogger.trace({ url }, "making request");

  /**
   * Note: We have to force convert the URL object back to a string here,
   * because many of our tests mock the global fetch call based on specific
   * url strings being passed in as the first argument. They all fail if an object
   * gets passed in instead (in this case a URL instance)
   */
  return undici.request(url.toString(), { headers }).then(async (r) => {
    // If there are headers, get the correlation ID. There may not be one, but
    // that's beside the point. We'll attach the correlation ID to downstream
    // log messages about the success/failure of this response.
    const correlationID = r.headers?.["x-correlation-id"];

    if (r.statusCode >= 200 && r.statusCode < 400) {
      fetchLogger.trace({ path, correlationID });

      // 204 means no content
      if (r.status === 204) {
        return new Promise((resolve) => resolve({}));
      }

      // Cache the value, and then return the JSON
      // response if there is a valid cache-control
      // header value in the response
      if (USE_REDIS) {
        const ttl = getTTLFromResponse(r);
        if (ttl) {
          const json = await r.body.json();
          await saveToRedis(url.pathname, json, ttl);
          return json;
        }
      }
      return r.body.json();
    }

    const response = await r.body.json();
    fetchLogger.error({
      path,
      status: r.status,
      correlationID,
      response,
    });

    // If there was a server error, retry. These are often temporary.
    if (r.statusCode >= 500) {
      const error = new Error();
      error.cause = { ...response, status: r.statusCode };
      return Promise.reject(error);
    }

    // For request errors, don't retry. They're not likely to resolve on their
    // own so there's no point.
    return { status: r.statusCode, ...response, error: true };
  });
};

export const fetchAPIJson = async (path, { wait = sleep } = {}) =>
  internalFetch(path)
    .catch((e) => {
      if (e.statusCode === 429) {
        return { ...e.cause, error: true, statusCode: e.statusCode };
      } else {
        return wait(75).then(() => internalFetch(path));
      }
    })
    .catch(() => wait(124).then(() => internalFetch(path)))
    .catch(() => wait(204).then(() => internalFetch(path)))
    .catch(() => wait(337).then(() => internalFetch(path)))
    .catch((e) => {
      if (e instanceof SyntaxError) {
        // this can happen if the API or proxy returns HTML
        fetchLogger.error({ err: e, path }, "invalid JSON");
      } else {
        fetchLogger.error({ err: e, path });
      }
      return { ...e.cause, error: true };
    });

export { BASE_URL, BASE_NWSCONNECT_URL, fetchAPIJson as default };

File: api-interop-layer/util/util-golang/fetch.go
package util_golang

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

var (
	BaseURL     = "https://api.weather.gov"
	BaseGhwoURL = "https://www.weather.gov"
	HTTPClient  = &http.Client{Timeout: 10 * time.Second} // Default client
	// Hook for sleeping in tests
	SleepFunc = time.Sleep
)

func init() {
	if v := os.Getenv("API_URL"); v != "" {
		BaseURL = v
	}
	if v := os.Getenv("GHWO_URL"); v != "" {
		BaseGhwoURL = v
	}
}

type FetchError struct {
	Status   int
	Response interface{}
	Err      error
}

func (e *FetchError) Error() string {
	return fmt.Sprintf("fetch error: status %d", e.Status)
}

// InternalFetch mirrors the internalFetch logic with retry support handling in the public wrapper
func internalFetch(path string) (interface{}, error) {
	targetURL := path
	if !strings.HasPrefix(path, "http") {
		// Handle relative path: join with BaseURL
		// JS: new URL(path, BASE_URL)
		// Go simple join
		u, _ := url.Parse(BaseURL)
		u.Path, _ = url.JoinPath(u.Path, path)
		// Note: url.JoinPath added in Go 1.19. user's go.mod says 1.21.
		targetURL = u.String()
	}

	u, err := url.Parse(targetURL)
	if err != nil {
		return nil, err
	}

	isGHWO := u.Hostname() == "www.weather.gov" && strings.HasPrefix(u.Path, "/source/")
	isAlert := strings.Contains(u.Path, "alerts")

	if isGHWO {
		// Switch to GHWO Base URL logic
		// JS: url = new URL(url.pathname, BASE_GHWO_URL);
		ghwo, _ := url.Parse(BaseGhwoURL)
		ghwo.Path = u.Path
		u = ghwo
	}

	// Redis Cache Check
	if UseRedis && !isGHWO && !isAlert {
		cached, err := GetFromRedis(u.Path)
		if err == nil && cached != "" {
			var result interface{}
			if err := json.Unmarshal([]byte(cached), &result); err == nil {
				return result, nil
			}
		}
	}

	req, err := http.NewRequest("GET", u.String(), nil)
	if err != nil {
		return nil, err
	}

	// Headers
	if apiKey := os.Getenv("API_KEY"); apiKey != "" {
		req.Header.Set("API-Key", apiKey)
	}
	// wx-host logic from JS
	originalU, _ := url.Parse(targetURL)
	if originalU.Host != "" {
		req.Header.Set("wx-host", originalU.Host)
	}

	resp, err := HTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// 200-299 Success
	if resp.StatusCode >= 200 && resp.StatusCode < 400 {
		// Cache Logic
		if UseRedis && !isGHWO && !isAlert {
			ttl := GetTTLFromResponse(resp.Header)
			if ttl > 0 {
				SaveToRedis(u.Path, string(bodyBytes), ttl)
			}
		}

		var result interface{}
		if err := json.Unmarshal(bodyBytes, &result); err != nil {
			// JS catches SyntaxError and returns { ...e.cause, error: true }
			// Here we return simple error object
			log.Printf("json parse error: %v", err)
			return map[string]interface{}{"error": true, "message": "invalid JSON"}, nil
		}
		return result, nil
	}

	// Error handling
	var responseData interface{}
	// Try to parse error body as JSON
	_ = json.Unmarshal(bodyBytes, &responseData)

	// Return error for 5xx to trigger retry
	if resp.StatusCode >= 500 {
		return nil, &FetchError{Status: resp.StatusCode, Response: responseData}
	}

	// 4xx returns error object but success (no retry)
	if respMap, ok := responseData.(map[string]interface{}); ok {
		respMap["error"] = true
		respMap["status"] = resp.StatusCode
		return respMap, nil
	}

	// Fallback
	return map[string]interface{}{
		"error":  true,
		"status": resp.StatusCode,
	}, nil

}

// FetchAPIJson fetches JSON from API with retries
func FetchAPIJson(path string) (interface{}, error) {
	delays := []time.Duration{
		75 * time.Millisecond,
		124 * time.Millisecond,
		204 * time.Millisecond,
		337 * time.Millisecond,
	}

	res, err := internalFetch(path)
	if err == nil {
		return res, nil
	}

	// If error is not a helper FetchError (e.g. network error) or is 5xx, we retry
	// If it was a 4xx, internalFetch returned result (not error), so we already returned above.

	for _, d := range delays {
		SleepFunc(d)
		res, err = internalFetch(path)
		if err == nil {
			return res, nil
		}
	}

	// After all retries
	// Return the last error structure if available
	if fErr, ok := err.(*FetchError); ok {
		// If it's a 5xx that failed retries
		// JS logic: returns { ...response, error: true }
		// Wait, JS: catches finally.
		/*
		  .catch((e) => {
		      if (e instanceof SyntaxError) { ... }
		      return { ...e.cause, error: true };
		    });
		*/
		// My internalFetch returns FetchError for 5xx.
		if respMap, ok := fErr.Response.(map[string]interface{}); ok {
			respMap["error"] = true
			return respMap, nil
		}
		return map[string]interface{}{"error": true, "status": fErr.Status}, nil
	}

	// Network error etc
	return map[string]interface{}{"error": true, "message": err.Error()}, nil
}
